In [8]:
# imports
import numpy             as np
import matplotlib.pyplot as plt
import pandas            as pd
import math
import keras
from scipy.stats  import norm
from keras.models import Sequential
from keras.layers import Dense, LSTM, Dropout
from keras        import backend as K
from datetime     import datetime
from keras.optimizers      import RMSprop
from sklearn.preprocessing import MinMaxScaler
from matplotlib.pylab      import rcParams
%matplotlib inline
rcParams['figure.figsize'] = 15, 6

# number of inputs to be fed
look_back    = 288
# number of outputs to be generated
look_forward = 24
# the number of stations
stations     = 66

# Prepares the Dataset For the Models

In [2]:
# The interval between each dataset (original data in 5 minute intervals)
time_grouping = '5min'
# load the dataset
dataframe = pd.read_csv('ignored_assets/paxout_table.csv', engine='python', nrows=288*21)
dataframe['time_bucket'] = pd.to_datetime(dataframe['time_bucket'])
dataframe = dataframe.set_index('time_bucket')
# dataframe['total'] = dataframe.sum(axis=1)
dataframe['day_hour'] = dataframe.index.round(time_grouping)
dataframe = dataframe.groupby('day_hour').sum()
# removes the timestamp at column 67
dataset_orig = dataframe.values[:,:stations]
dataset_orig = dataset_orig.astype('float32')
# scale the dataset
scaler = MinMaxScaler(feature_range=(0, 1))
dataset = scaler.fit_transform(dataset_orig)

# convert an array of values into a dataset matrix, adjusted to make a dateset that is 66 wide
def create_dataset(dataset, look_back=1, look_forward=2):
    dataX, dataY = [], []
    np.array(dataY)
    for i in range(len(dataset)-look_back-1):
        a = dataset[i:(i+look_back),:]
        # Makes sure that the array isn't the last 2 or 3 or whatever bits. It has to be the full 24
        if len(dataset[i + look_back:i+look_back+look_forward, 0]) == look_forward:
            dataX.append(a.T)
            dataY.append(dataset[i + look_back:i+look_back+look_forward, :].T)
    return np.array(dataX), np.array(dataY)

# split into train and test sets
train_size = 288*14
test_size = len(dataset) - train_size
train, test = dataset[0:train_size,:], dataset[train_size:len(dataset)-look_forward,:]

# reshape into X=[t, t-1, t-2,..., t-look_back] and Y=[t+1, t+2,... t+look_forward]
trainX, trainY = create_dataset(train, look_back, look_forward)
trainX, trainY = create_dataset(test, look_back, look_forward)

# The Loss Functions

In [3]:
def top_heavy_loss (y_true, y_pred):
    w = np.arange(1,0, -1./24) ** 2
    w = (w / w.sum())[:,None]
    W = K.variable(value = w)
    return K.dot( K.abs(y_pred-y_true), W)

def top_heavy_tail_heavy (y_true, y_pred):
    # This is the normal weights that decrease
    # w = np.arange(1, 0, -1./24)**1.3
    # This is the Weights with one difference, The last is weighted as equally as the first
    w = np.append(np.arange(1,1./23, -1./24)**2.5, [1])
    w = w / w.sum()
    w = w[:,None]
    W = K.variable(value = w)
    return K.dot( K.abs(y_pred-y_true), W)


# The Custom Grid Search Function

In [4]:
'''
  My Custom Grid Search Function:
    * trainX       - The input data for the model
    * trainY       - The target data for the model
    * create_model - The function which returns a compiled model
    * param_grid   - The parameters for the function to grid search through
    * batch_size   - The default batch size if not otherwise specified directly or as a parameter to grid search
    * epochs       - The default number of epochs if not otherwise specified directly 
                     or as a parameter to grid search through
    * verbose      - The verbose level when training the individual prameters
    * seed         - The seed to be applied to each model 
'''
def custom_grid_search(trainX, trainY, create_model, param_grid, batch_size=25, epochs=10, verbose=3, seed=None):
    # creates an array of the keys
    keys = np.array(param_grid.keys())
    # feeds the keys, the full dictionary, and a fresh dictionary to the recursive function
    # to get an array of dictionaries of every possible combination of parameters
    indiv_params = combine_parameters (keys, param_grid, dict())
    # An array which will keep track of the accuracy of each iteration of the parameters
    results = np.zeros(indiv_params.size)
    # Runs each of the possible combinations
    for i in range(indiv_params.size):
        # prevents changing the original set of parameters
        indiv_param = indiv_params[i].copy()
        # Saves batch_size and epochs for the fitting of the model, not the creation/compilation of
        if 'batch_size' in keys:
            batch_size = indiv_param.pop('batch_size', None)
        if 'epochs' in keys:
            epochs = indiv_param.pop('epochs', None)
        # if a seed has been specified, use it
        if seed != None:
            np.random.seed(seed)
        # Creates the model and trains it
        model = create_model(**indiv_param)
        model.fit(trainX, trainY, epochs=epochs, batch_size=batch_size, verbose=verbose)
        # Stores this iteration's results
        results[i] = np.mean(model.evaluate(trainX, trainY, verbose=3))
        # Clears the backend session in an attempt to save memory with each model
        K.clear_session()
    # finds the index of the result with the lowest loss value
    best_i = np.argwhere(results == min(results) )
    
    best_results = results[best_i].reshape(1)[0]
    best_params  = indiv_params[best_i].reshape(1)[0]
    # Returns all the results and their respective parameters in order
    return dict(results=results,
                parameters=indiv_params,
                best_results = best_results,
                best_parameters = best_params)

'''
  Given a Dictionary, the keys yet to be looped through, and a specific instance of the dictionary 
  it returns an array of dictionaries showing all unique combinations of the values
  
  e.g. given a dictionary (a=[0, 1], b=[0, 1]), it will return the following array
  array([
    dict(a=0, b=0),
    dict(a=0, b=1),
    dict(a=1, b=0),
    dict(a=1, b-1)
  ])
  
  The Structure is as follows:

    The function is given an array of keys ['a', 'b'] and a dictionary (a=[0,1], b=[0, 1])

    It loops through the first key, in this case a ∈ {0, 1} and removes the first key from the array,
    its new value being ['b']

    It then loops through all the values of the first key, creating a variable called specific_dict, 
    which is a dictionary with the given value of a (as well as any previously defined keys in specific_dict)
      (a=0), and (a=1)

    It then calls itself, providing the updated list of keys yet to be iterated through, the full dictionary,
    and the specific_dict each for loop

    Once the list of keys yet to be iterated through is of size 1, it ceases the recursion
  
'''
def combine_parameters(keys, full_dict, specific_dict):
    result = np.array([])
    # If the size of the keys array is 1, that means this is the final key to be looped through, cease recursion
    if keys.size == 1:
        for i in range(len(full_dict[keys[0]])):
            specific_dict = specific_dict.copy()
            # the result will be the specific dict followed by every value in this 
            specific_dict[keys[0]] = full_dict[keys[0]][i]
            result = np.append(result, specific_dict)
        return result
    else:
        # keeps the first key
        current_key = keys[0]
        # removes the first key from the list of keys that havent been looped through yet
        keys = keys[1:]
        # for every value of the current_key
        for i in range(len(full_dict[current_key])):
            specific_dict = specific_dict.copy()
            specific_dict[current_key] = full_dict[current_key][i]
            result = np.append(result, combine_parameters(keys, full_dict, specific_dict))
    return result

In [5]:

# creates a model but does not compile the model
def create_model(neurons, layers):
    dropout_rate      = 0.3
    optimizer=RMSprop(lr=0.0001, rho=0.9, epsilon=1e-08, decay=0.0)
    model = Sequential()
    model.add(Dropout(dropout_rate, input_shape=(stations, look_back)))
    lstm = LSTM(256, recurrent_dropout=dropout_rate, return_sequences=True)
    model.add(lstm)
    for i in range(layers):
        model.add(Dropout(dropout_rate))
        model.add(Dense(neurons))
    model.add(Dense(look_forward))
    model.compile(loss=top_heavy_loss, optimizer=optimizer, metrics=[top_heavy_loss])
    return model

# The Parameters to Search Through
layers = [3, 4, 5]
neurons = [32, 64, 128, 256]

# Defines it as a dictionary
param_grid = dict(layers = layers,
                  neurons = neurons)

# Begins the Search For the Optimal Properties
grid_result = custom_grid_search(trainX, trainY, create_model, batch_size=20, epochs=30, param_grid=param_grid)

K.clear_session()

# summarize results
print("Best: %f using %s" % (grid_result['best_results'], str(grid_result['best_parameters'])))

loss = grid_result['results']
params = grid_result['parameters']

for l, p in zip(loss, params):
    print l, p

Epoch 1/30
Epoch 2/30
Epoch 3/30
Epoch 4/30
Epoch 5/30
Epoch 6/30
Epoch 7/30
Epoch 8/30
Epoch 9/30
Epoch 10/30
Epoch 11/30
Epoch 12/30
Epoch 13/30
Epoch 14/30
Epoch 15/30
Epoch 16/30
Epoch 17/30
Epoch 18/30
Epoch 19/30
Epoch 20/30
Epoch 21/30
Epoch 22/30
Epoch 23/30
Epoch 24/30
Epoch 25/30
Epoch 26/30
Epoch 27/30
Epoch 28/30
Epoch 29/30
Epoch 30/30
Epoch 1/30
Epoch 2/30
Epoch 3/30
Epoch 4/30
Epoch 5/30
Epoch 6/30
Epoch 7/30
Epoch 8/30
Epoch 9/30
Epoch 10/30
Epoch 11/30
Epoch 12/30
Epoch 13/30
Epoch 14/30
Epoch 15/30
Epoch 16/30
Epoch 17/30
Epoch 18/30
Epoch 19/30
Epoch 20/30
Epoch 21/30
Epoch 22/30
Epoch 23/30
Epoch 24/30
Epoch 25/30
Epoch 26/30
Epoch 27/30
Epoch 28/30
Epoch 29/30
Epoch 30/30
Epoch 1/30
Epoch 2/30
Epoch 3/30
Epoch 4/30
Epoch 5/30
Epoch 6/30
Epoch 7/30
Epoch 8/30
Epoch 9/30
Epoch 10/30
Epoch 11/30
Epoch 12/30
Epoch 13/30
Epoch 14/30
Epoch 15/30
Epoch 16/30
Epoch 17/30
Epoch 18/30
Epoch 19/30
Epoch 20/30
Epoch 21/30
Epoch 22/30
Epoch 23/30
Epoch 24/30
Epoch 25/30
Epoch 2

In [6]:
K.clear_session()

In [ ]:
'''
Best: 0.064527 using {'layers': 4, 'neurons': 128}
0.0673357171075 {'layers': 3, 'neurons': 128}
0.0656038054987 {'layers': 3, 'neurons': 256}
0.0661557383374 {'layers': 3, 'neurons': 128}
0.0645914614857 {'layers': 4, 'neurons': 128}
0.0649715498542 {'layers': 4, 'neurons': 256}
0.0645265403596 {'layers': 4, 'neurons': 128}
0.0653111206801 {'layers': 5, 'neurons': 128}
0.0680737998237 {'layers': 5, 'neurons': 256}
0.06743119954 {'layers': 5, 'neurons': 128}
0.069117619517 {'layers': 6, 'neurons': 128}
0.0711870008325 {'layers': 6, 'neurons': 256}
0.0663055349285 {'layers': 6, 'neurons': 128}


Best: 0.064348 using {'layers': 3, 'neurons': 64}
0.0651559979344 {'layers': 3, 'neurons': 32}
0.0643481532891 {'layers': 3, 'neurons': 64}
0.064456228497 {'layers': 3, 'neurons': 128}
0.065554653921 {'layers': 3, 'neurons': 256}
0.0681183533893 {'layers': 4, 'neurons': 32}
0.0654155373649 {'layers': 4, 'neurons': 64}
0.0668186072475 {'layers': 4, 'neurons': 128}
0.0724844430473 {'layers': 4, 'neurons': 256}
0.0704560831792 {'layers': 5, 'neurons': 32}
0.0654539013285 {'layers': 5, 'neurons': 64}
0.0650072880311 {'layers': 5, 'neurons': 128}
0.0648792156731 {'layers': 5, 'neurons': 256}


Best: 0.064915 using {'layers': 3, 'neurons': 64}
0.0650385718821 {'layers': 3, 'neurons': 32}
0.0649145629605 {'layers': 3, 'neurons': 64}
0.0649556586611 {'layers': 3, 'neurons': 128}
0.068413673628 {'layers': 3, 'neurons': 256}
0.0674589839805 {'layers': 4, 'neurons': 32}
0.0657898077097 {'layers': 4, 'neurons': 64}
0.0698334501545 {'layers': 4, 'neurons': 128}
0.0652247917623 {'layers': 4, 'neurons': 256}
0.0680475525851 {'layers': 5, 'neurons': 32}
0.0663156365839 {'layers': 5, 'neurons': 64}
0.076172077276 {'layers': 5, 'neurons': 128}
0.0676428279875 {'layers': 5, 'neurons': 256}

'''

In [38]:

# creates a model but does not compile the model
def create_model(dropout_rate = 0.3):
    layers            = 3
    neurons           = 128
    model = Sequential()
    model.add(Dropout(dropout_rate, input_shape=(stations, look_back)))
    lstm = LSTM(256, recurrent_dropout=dropout_rate, return_sequences=True)
    model.add(lstm)
    for i in range(layers):
        model.add(Dropout(dropout_rate))
        model.add(Dense(neurons))
    model.add(Dense(look_forward))
    model.compile(loss=top_heavy_loss, optimizer=RMSprop(lr=0.0001, rho=0.9, epsilon=1e-08, decay=0.0), metrics=[top_heavy_loss])
    return model

# The Parameters to Search Through
dropout_rate = [0.175, 0.2, 0.225, 0.250, 0.275, 0.300, 0.325, 0.35]

# Defines it as a dictionary
param_grid = dict(dropout_rate=dropout_rate)

# The number of times to repeat the grid search for the average
num_iterations = 3
# this will store three grid_result's
grid_results = np.zeros(num_iterations).astype('object')

# creates a m x n matrix where m is the number of grid searches and n is the number of unique shit
grid_losses = np.zeros((num_iterations, len(dropout_rate) ))

for i in range(grid_results.size):
    # Begins the Search For the Optimal Properties
    grid_result = custom_grid_search(trainX, trainY, create_model, epochs=30, batch_size=30, param_grid=param_grid)
    # You can never be too careful
    K.clear_session()
    # stores the entire dictionary into a variable
    grid_results[i] = grid_result
    # stores the results of the parameters into the matrix
    grid_losses[i,:] = grid_result['results']
    
# averages the losses of each of the runs
aver = grid_losses.mean(axis=0)

# prints all of the results
for i in grid_results:
    loss = grid_results[i]['results']
    params = grid_results[i]['parameters']
    for l, p in zip(loss, params):
        print l, p

# summarize results
print("Best: %f using %s" % (aver.min(), str(grid_results['parameters'][np.argwhere(aver == aver.min())]  )))

Epoch 1/30
Epoch 2/30
Epoch 3/30
Epoch 4/30
Epoch 5/30
Epoch 6/30
Epoch 7/30
Epoch 8/30
Epoch 9/30
Epoch 10/30
Epoch 11/30
Epoch 12/30
Epoch 13/30
Epoch 14/30
Epoch 15/30
Epoch 16/30
Epoch 17/30
Epoch 18/30
Epoch 19/30
Epoch 20/30
Epoch 21/30
Epoch 22/30
Epoch 23/30
Epoch 24/30
Epoch 25/30
Epoch 26/30
Epoch 27/30
Epoch 28/30
Epoch 29/30
Epoch 30/30
Epoch 1/30
Epoch 2/30
Epoch 3/30
Epoch 4/30
Epoch 5/30
Epoch 6/30
Epoch 7/30
Epoch 8/30
Epoch 9/30
Epoch 10/30
Epoch 11/30
Epoch 12/30
Epoch 13/30
Epoch 14/30
Epoch 15/30
Epoch 16/30
Epoch 17/30
Epoch 18/30
Epoch 19/30
Epoch 20/30
Epoch 21/30
Epoch 22/30
Epoch 23/30
Epoch 24/30
Epoch 25/30
Epoch 26/30
Epoch 27/30
Epoch 28/30
Epoch 29/30
Epoch 30/30
Epoch 1/30
Epoch 2/30
Epoch 3/30
Epoch 4/30
Epoch 5/30
Epoch 6/30
Epoch 7/30
Epoch 8/30
Epoch 9/30
Epoch 10/30
Epoch 11/30
Epoch 12/30
Epoch 13/30
Epoch 14/30
Epoch 15/30
Epoch 16/30
Epoch 17/30
Epoch 18/30
Epoch 19/30
Epoch 20/30
Epoch 21/30
Epoch 22/30
Epoch 23/30
Epoch 24/30
Epoch 25/30
Epoch 2

Epoch 12/30
Epoch 13/30
Epoch 14/30
Epoch 15/30
Epoch 16/30
Epoch 17/30
Epoch 18/30
Epoch 19/30
Epoch 20/30
Epoch 21/30
Epoch 22/30
Epoch 23/30
Epoch 24/30
Epoch 25/30
Epoch 26/30
Epoch 27/30
Epoch 28/30
Epoch 29/30
Epoch 30/30


IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

In [10]:

loss = grid_result['results']
params = grid_result['parameters']

for l, p in zip(loss, params):
    print l, p

0.0618390910358 {'dropout_rate': 0.2}
0.0650889443135 {'dropout_rate': 0.25}
0.0635920508872 {'dropout_rate': 0.3}
0.0633840294026 {'dropout_rate': 0.35}
0.065912780799 {'dropout_rate': 0.4}



Best: 0.061839 using {'dropout_rate': 0.2}
0.0618390910358 {'dropout_rate': 0.2}
0.0650889443135 {'dropout_rate': 0.25}
0.0635920508872 {'dropout_rate': 0.3}
0.0633840294026 {'dropout_rate': 0.35}
0.065912780799 {'dropout_rate': 0.4}

Best: 0.064722 using {'dropout_rate': 0.25}
0.0659280873121 {'dropout_rate': 0.2}
0.0647218886215 {'dropout_rate': 0.25}
0.065688662098 {'dropout_rate': 0.3}
0.0675317480925 {'dropout_rate': 0.35}
0.0655382096608 {'dropout_rate': 0.4}

In [24]:
for i in a:
    print i

{'a': 1}
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0


In [25]:
a = np.array([[1,2,3],[1,2,3]])

In [29]:
a = np.arange(4)

In [34]:
np.argwhere(a == 0)

array([[0]])

In [37]:
K.clear_session()